model.py

```diff
@@ -6616,6 +6618,126 @@ def update_availability(
+            # This is an old event and its effect on availability has
+            # already been taken into account.
+            return
+
+        if self.last_checked and event_date == CirculationEvent.NO_DATE:
+            # We have a history for this LicensePool and we don't know
+            # where this event fits into that history. Ignore the
+            # event.
+            return
+
+        (new_licenses_owned, new_licenses_available, 
+         new_licenses_reserved, 
+         new_patrons_in_hold_queue) = self._calculate_change_from_one_event(
+             event_type, delta
+         )
+
+        changes_made = self.update_availability(
+            new_licenses_owned, new_licenses_available, 
+            new_licenses_reserved, new_patrons_in_hold_queue,
+            analytics=analytics, as_of=event_date
+        )
+        if not changes_made:
+            # Even if nothing about availability actually changed, we
+            # want to record the fact that the event happened.
+            self.collect_analytics_event(
+                analytics, event_type, event_date, 0, 0
+            )
+
+    def _calculate_change_from_one_event(self, type, delta):
+        new_licenses_owned = self.licenses_owned
+        new_licenses_available = self.licenses_available
+        new_licenses_reserved = self.licenses_reserved
+        new_patrons_in_hold_queue = self.patrons_in_hold_queue
+
+        def deduct(value):
+            # It's impossible for any of these numbers to be
+            # negative.
+            return max(value-delta, 0)
+
+        CE = CirculationEvent
+        added = False
+        if type == CE.DISTRIBUTOR_HOLD_PLACE:
+            new_patrons_in_hold_queue += delta
+            if new_licenses_available:
+                # If someone has put a book on hold, it must not be
+                # immediately available.
+                new_licenses_available = 0
+        elif type == CE.DISTRIBUTOR_HOLD_RELEASE:
+            new_patrons_in_hold_queue = deduct(new_patrons_in_hold_queue)
+        elif type == CE.DISTRIBUTOR_CHECKIN:
+            if self.patrons_in_hold_queue == 0:
+                new_licenses_available += delta
+            else:
+                # When there are patrons in the hold queue, checking
+                # in a single book does not make new licenses
+                # available.  Checking in more books than there are
+                # patrons in the hold queue _does_ make books
+                # available.  However, in neither case do patrons
+                # leave the hold queue. That will happen in the near
+                # future as DISTRIBUTOR_AVAILABILITY_NOTIFICATION events 
+                # are sent out.
+                if delta > new_patrons_in_hold_queue:
+                    new_licenses_available += (delta-new_patrons_in_hold_queue)
+        elif type == CE.DISTRIBUTOR_CHECKOUT:
+            new_licenses_available = deduct(new_licenses_available)
```

Comment:
Is there a case when this changes licenses_reserved instead of licenses_available?